### PR TITLE
Connector cli: command to clean old Nango connections

### PR DESCRIPTION
**Pull request by PopDaph**
Part of: https://github.com/dust-tt/tasks/issues/258

Cli command to remove the connectionId if we don't have a matching `SlackConfiguration` in db. 

Note: Nango does not offer a way to filter the list of connections: https://docs.nango.dev/api-reference/connection/list, so I'm manually filtering them for Slack. This is probably not sustainable on the long term, and I'm not even sure this will be okay with the current number of connections but let's try and find a plan B if needed.

### CODE REVIEW

File: connectors/src/admin/cli.ts [hunk 1]
```diff
@@ -383,6 +391,61 @@ const slack = async (command: string, args: parseArgs.ParsedArgs) => {
...
+        if (!slackConfigurations.find((sc) => sc.slackTeamId === slackTeamId)) {
+          const answer: string = await askQuestion(
+            `Do you want to delete connection ${connection.connection_id} for team ${slackTeamId}? (y/N) `
+          );
+          if (answer.toLowerCase() !== "y") {
+            continue;
+          }
+
+          console.log("Deleting connection...");
+          await nangoDeleteConnection(
```
Comment (spolu) [inline connectors/src/admin/cli.ts:441]: This won't revoke the token will it? We probably want to revoke the token as well. We can have it for slack only and error otherwise (and not delete the nango connection if that's the case for other platforms)

File: connectors/src/admin/cli.ts [hunk 2]
```diff
@@ -17,8 +18,15 @@ import { launchSlackSyncOneThreadWorkflow } from "@connectors/connectors/slack/t
...
 import { GoogleDriveFiles } from "@connectors/lib/models/google_drive";
 import { NotionDatabase, NotionPage } from "@connectors/lib/models/notion";
+import { SlackConfiguration } from "@connectors/lib/models/slack";
+import {
+  nango_client,
+  nangoDeleteConnection,
+} from "@connectors/lib/nango_client";
 import { Result } from "@connectors/lib/result";
 
+const { NANGO_SLACK_CONNECTOR_ID = "" } = process.env;
```
Comment (lasryaric) [inline connectors/src/admin/cli.ts:28]: I think we should error if it's undefined at the time we need it and not rely on a default.

File: connectors/src/admin/cli.ts (hunk 1 above)
Comment (lasryaric) [reply connectors/src/admin/cli.ts:441]: To revoke a Slack token I think we need to call this api endpont:
https://api.slack.com/methods/auth.revoke

We should not uninstall because if the user logged in with another token, it will uninstall our bot from their workspace (as well as revoke the token).

File: connectors/src/admin/cli.ts [hunk 3]
```diff
@@ -383,6 +391,61 @@ const slack = async (command: string, args: parseArgs.ParsedArgs) => {
 
       break;
     }
+
+    case "clean-nango-connection-id": {
+      const slackConfigurations = await SlackConfiguration.findAll();
+      const connections = await nango_client().listConnections();
+      const slackConnections = connections.connections.filter(
+        (connection: {
```
Comment (lasryaric) [inline connectors/src/admin/cli.ts:399]: We also need to filter on the creation date, we can't revoke token that are younger than I would say one hour (to be very safe).
We could revoke a token that has just been created on Nango but not yet in connectors if we don't do that. (on top of the discussed edge case of replication lag latency).

**Review by lasryaric — CHANGES_REQUESTED** [review]
Thanks @PopDaph. To be tested meticulously.
I left a few comments.

File: connectors/src/admin/cli.ts (hunk 1 above)
Comment (PopDaph) [reply connectors/src/admin/cli.ts]: I called our new method `uninstallSlack` instead

File: connectors/src/admin/cli.ts (hunk 2 above)
Comment (PopDaph) [reply connectors/src/admin/cli.ts:26]: Done!

File: connectors/src/admin/cli.ts (hunk 3 above)
Comment (PopDaph) [reply connectors/src/admin/cli.ts:405]: Done

File: connectors/src/lib/nango_client.ts [hunk 4]
```diff
@@ -48,6 +48,10 @@ class CustomNango extends Nango {
       throw e;
     }
   }
+
+  async listConnections() {
+    return await super.listConnections();
```
Comment (lasryaric) [inline connectors/src/lib/nango_client.ts:53]: Why is that needed if we only call super.listConnections()?

**Review by spolu — DISMISSED** [review]
LGTM

If I understand correctly it takes all connections (more than 1h old) that have no slack configurations associated and propose to delete them?

File: connectors/src/admin/cli.ts [hunk 5]
```diff
@@ -383,6 +389,71 @@ const slack = async (command: string, args: parseArgs.ParsedArgs) => {
...
+        });
+      };
+
+      for (const connection of slackConnections) {
+        connectionDetail = await nango_client().getConnection(
+          connection.provider,
+          connection.connection_id
+        );
+
+        slackTeamId = connectionDetail.credentials.raw.team.id;
```
Comment (lasryaric) [inline connectors/src/admin/cli.ts:442]: Why is slackTeamId declared in the scope above the for loop? It looks like it's not used outside of this scope.

Comment (lasryaric) [reply connectors/src/admin/cli.ts:442]: Same thing for `connectionDetail`

**Review by lasryaric — CHANGES_REQUESTED** [review]
Left two other questions.

**Comment by lasryaric** [comment]
Also I asked about paginating this endpoint on their Slack community channel.
Conversation is here: https://nango-community.slack.com/archives/C04ABR352H0/p1701095836255929

**Comment by spolu** [comment]
Agreed on higher priority of conversation rendering :+1:

File: connectors/src/admin/cli.ts (hunk 5 above)
Comment (PopDaph) [reply connectors/src/admin/cli.ts:442]: Only to reuse the same variable throughout the loop instead of redefining new consts.

File: connectors/src/lib/nango_client.ts (hunk 4 above)
Comment (PopDaph) [reply connectors/src/lib/nango_client.ts:53]: Because `nango_client()` that is used everywhere is CustomNango, and I think it's cleaner to reuse the same pattern as everywhere else 🙏🏻

File: connectors/src/admin/cli.ts (hunk 5 above)
Comment (lasryaric) [reply connectors/src/admin/cli.ts:442]: I think this should be a const and only creates room for error. Can we move it to a const?

**Comment by PopDaph** [comment]
@lasryaric I answered your questions, I think code should be shipped as is for both 🙏🏻

File: connectors/src/admin/cli.ts (hunk 5 above)
Comment (lasryaric) [reply connectors/src/admin/cli.ts:442]: Sorry, to be clearer: here we run into the potential of re-using a previous value of a non const variable where there is no reason to not use a const.
So lets use a const to limit the error surface. Especially in a code that is so sensitive and risky.

**Review by lasryaric — APPROVED** [review]
LGTM. Thanks @PopDaph !

File: connectors/src/admin/cli.ts (hunk 5 above)
Comment (PopDaph) [reply connectors/src/admin/cli.ts]: I don't think this was dangerous and that this is necessarily a bad pattern, but since you insisted I added a commit to redefine consts inside the loop

**Comment by PopDaph** [comment]
(Merging but won't run it until I shipped the Production check)
